File: pegasusX/apps/payload-app-ios/payload-app-ios/Utilities/ManifestBoard.swift
import Foundation

enum ManifestBoard {
    static let states = ["DRAFT", "LOADING", "SEALED", "DISPATCHED"]

    static func canonicalState(_ state: String?) -> String {
        let s = (state ?? "").trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
        return Self.states.contains(s) ? s : ""
    }

    static func isBoardState(_ state: String?) -> Bool {
        !canonicalState(state).isEmpty
    }

    struct Column: Equatable {
        let state: String
        let trucks: [Truck]
    }

    static func attach(trucks: [Truck], manifests: [Manifest]) -> [Truck] {
        trucks.map { truck in
            if !canonicalState(truck.truckStatus).isEmpty { return truck }
            let match = manifests
                .filter { $0.matchesTruck(truck.id) && isBoardState($0.state) }
                .max(by: { ($0.createdAt ?? "") < ($1.createdAt ?? "") })
            guard let match else { return truck }
            return truck.withBoard(
                status: canonicalState(match.state),
                used: match.totalVolumeVu,
                max: match.maxVolumeVu,
                stops: match.stopCount
            )
        }
    }

    static func group(_ trucks: [Truck]) -> [Column] {
        states.map { state in
            Column(state: state, trucks: trucks.filter { canonicalState($0.truckStatus) == state })
        }
    }

    static func unassigned(_ trucks: [Truck]) -> [Truck] {
        trucks.filter { canonicalState($0.truckStatus).isEmpty }
    }
}

File: pegasusX/apps/payload-app-android/app/src/main/java/com/pegasus/payload/ui/home/ManifestBoard.kt
package com.pegasus.payload.ui.home

import com.pegasus.payload.data.model.Manifest
import com.pegasus.payload.data.model.Truck

/** GS-U7 payload board: columns are manifest states, not a single trucks count. */
object ManifestBoard {
    val BOARD_STATES = listOf("DRAFT", "LOADING", "SEALED", "DISPATCHED")

    fun canonicalState(state: String?): String {
        val s = state?.trim()?.uppercase().orEmpty()
        return if (s in BOARD_STATES) s else ""
    }

    fun isBoardState(state: String?): Boolean = canonicalState(state).isNotEmpty()

    data class Column(val state: String, val trucks: List<Truck>)

    fun attach(trucks: List<Truck>, manifests: List<Manifest>): List<Truck> {
        return trucks.map { truck ->
            if (canonicalState(truck.truckStatus).isNotEmpty()) return@map truck
            val match = manifests
                .filter { it.matchesTruck(truck.id) && isBoardState(it.state) }
                .maxByOrNull { it.createdAt }
            if (match == null) truck
            else truck.copy(
                truckStatus = canonicalState(match.state),
                usedVolumeVu = match.totalVolumeVu.toLong(),
                maxVolumeVu = match.maxVolumeVu.toLong(),
                stopCount = match.stopCount,
            )
        }
    }

    fun group(trucks: List<Truck>): List<Column> {
        return BOARD_STATES.map { state ->
            Column(state = state, trucks = trucks.filter { canonicalState(it.truckStatus) == state })
        }
    }

    fun unassigned(trucks: List<Truck>): List<Truck> =
        trucks.filter { canonicalState(it.truckStatus).isEmpty() }
}
